src/videoCache.js

```diff
@@ -0,0 +1,213 @@
+/**
+ * This module interacts with the server used to cache video ad content to be restored later.
+ * At a high level, the expected workflow goes like this:
+ *
+ *   - Request video ads from Bidders
+ *   - Generate IDs for each valid bid, and cache the key/value pair on the server.
+ *   - Return these IDs so that publishers can use them to fetch the bids later.
+ *
+ * This trickery helps integrate with ad servers, which set character limits on request params.
+ */
+
+import { ajax } from './ajax';
+import { tryAppendQueryString } from './utils';
+
+const PUT_URL = 'https://prebid.adnxs.com/pbc/v1/put'
+const GET_URL = 'https://prebid.adnxs.com/pbc/v1/get?';
+const EMPTY_VAST_RESPONSE = '<VAST version="3.0"></VAST>'
+
+/**
+ * These are the properties required on a Bid in order to cache and retrieve it.
+ *
+ * @typedef {object} CacheableBid
+ * @property {string} vastUrl A URL which loads some valid VAST XML.
+ */
+
+/**
+ * Function which wraps a URI that serves VAST XML, so that it can be loaded.
+ *
+ * @param {string} uri The URI where the VAST content can be found.
+ * @return A VAST URL which loads XML from the given URI.
+ */
+function wrapURI(uri) {
+  // Technically, this is vulnerable to cross-script injection by sketchy vastUrl bids.
+  // We could make sure it's a valid URI... but since we're loading VAST XML from the
+  // URL they provide anyway, that's probably not a big deal.
+  return '<VAST version="2.0"><Ad id=""><Wrapper><AdSystem>prebid.org wrapper</AdSystem><VASTAdTagURI><![CDATA[' +
+    uri +
+    ']]></VASTAdTagURI><Impression></Impression><Creatives></Creatives></Wrapper></Ad></VAST>'
+}
+
+/**
+ * Wraps a bid in the format expected by the prebid-server endpoints, or returns null if
+ * the bid can't be converted cleanly.
+ *
+ * @param {CacheableBid} bid
+ */
+function toStorageRequest(bid) {
+  return {
+    'value': wrapURI(bid.vastUrl)
+  };
+}
+
+/**
+ * If the server responded successfully, then it gave us a bunch of objects with uuid properties.
+ * Abstract our background dependency away by renaming them to cacheId.
+ */
+function fromStorageResponse(response) {
+  return {
+    cacheId: response.uuid
+  };
+}
+
+/**
+ * A function which should be called with the results of the storage operation.
+ *
+ * @callback videoCacheStoreCallback
+ *
+ * @param {Error} [error] The error, if one occurred.
+ * @param {?string[]} uuids An array of unique IDs. The array will have one element for each bid we were asked
+ *   to store. It may include null elements if some of the bids were malformed, or an error occurred.
```

Comment:
I wanted to do this... but actually couldn't come up with a clean way. It doesn't really have anything to do with server behavior.

My assumption: A caller who **knows** their objects are invalid will filter them out before calling this method.
 
Now imagine a caller with an **unknown** invalid object, who's trying to associate the IDs with their objects:

```
cache.store([validThing1, invalidThing, validThing2], function(uuids) {
  validThing1.uuid = uuids[0];
  invalidThing.uuid = uuids[1];
  validThing2.uuid = uuids[2]; 
});
```

So... at the API level, what are my options for preventing "garbage in"? These are the things I thought of:

1. Cache nothing, and reject any request with at least one invalid object 
2. Return a length-2 array with the valid UUIDs
3. Null out the indices of invalid requests, and maintain the indices for valid ones

(1) is a solid API...but seemed unnecessarily unforgiving.
(2) is error-prone, and produces subtle bugs in the calling code.
(3) is more forgiving, and leaves un-cached objects with an undefined ID.

In other languages, I'd do a `Map[Object, String]`... but I don't think javascript supports those in our browsers. It might be polyfill-able though? Anyway:

I went with (3) because it seemed most reasonable.
I'm open to (1) if you think it's better.
I'm strongly opposed to (2).

If you have anything else in mind that I didn't think of, let me know so we can run the pros/cons on them.